File: java/Samples/Biometrics/CS/SimpleVoicesSampleCS/EnrollFromFile.cs
using System;
using System.IO;
using System.Linq;
using System.Windows.Forms;
using Neurotec.Biometrics;
using Neurotec.Biometrics.Client;

namespace Neurotec.Samples
{
	public partial class EnrollFromFile : UserControl
	{
		#region Public constructor

		public EnrollFromFile()
		{
			InitializeComponent();
		}

		#endregion

		#region Private fields

		private NBiometricClient _biometricClient;
		private NSubject _subject;
		private NVoice _voice;

		private bool _defaultExtractDependentFeatures;
		private bool _defaultExtractIndependentFeatures;

		#endregion

		#region Public properties

		public NBiometricClient BiometricClient
		{
			get { return _biometricClient; }
			set
			{
				_biometricClient = value;
				_defaultExtractDependentFeatures = _biometricClient.VoicesExtractTextDependentFeatures;
				_defaultExtractIndependentFeatures = _biometricClient.VoicesExtractTextIndependentFeatures;
				chbExtractTextDependent.Checked = _defaultExtractDependentFeatures;
				chbExtractTextIndependent.Checked = _defaultExtractIndependentFeatures;
			}
		}

		#endregion

		#region Private methods

		private void DisableControls()
		{
			lblSoundFile.Text = string.Empty;
			lblStatus.Text = string.Empty;
			btnExtract.Enabled = false;
			btnSaveTemplate.Enabled = false;
			btnSaveVoice.Enabled = false;
		}

		private void OnExtractionCompleted(IAsyncResult r)
		{
			if (InvokeRequired)
			{
				BeginInvoke(new AsyncCallback(OnExtractionCompleted), r);
			}
			else
			{
				NBiometricTask task = _biometricClient.EndPerformTask(r);
				NBiometricStatus status = task.Status;
				if (status == NBiometricStatus.Ok)
				{
					btnSaveTemplate.Enabled = true;
					btnSaveVoice.Enabled = true;
					lblStatus.Text = @"Template extracted";
				}
				else
				{
					lblStatus.Text = string.Format("Extraction failed: {0}.", status);
				}
			}
		}

		#endregion

		#region Private form events

		private void EnrollFromFileLoad(object sender, EventArgs e)
		{
			DisableControls();
		}

		private void BtnOpenClick(object sender, EventArgs e)
		{
			_subject = null;
			_voice = null;

			DisableControls();

			if (openFileDialog.ShowDialog() == DialogResult.OK)
			{
				// Create a subject with voice record
				_voice = new NVoice { FileName = openFileDialog.FileName };
				_subject = new NSubject();
				_subject.Voices.Add(_voice);

				lblSoundFile.Text = openFileDialog.FileName;
				btnExtract.Enabled = true;
			}
		}

		private void BtnExtractClick(object sender, EventArgs e)
		{
			btnSaveVoice.Enabled = false;
			btnSaveTemplate.Enabled = false;
			lblStatus.Text = string.Empty;

			if (!chbExtractTextDependent.Checked && !chbExtractTextIndependent.Checked)
			{
				MessageBox.Show(@"No features configured to extract", @"Invalid options", MessageBoxButtons.OK, MessageBoxIcon.Warning);
				return;
			}

			try
			{
				_voice.PhraseId = Convert.ToInt32(nudPhraseId.Value);
				// Do voice extraction and segment voice from audio
				NBiometricTask task = _biometricClient.CreateTask(NBiometricOperations.Segment | NBiometricOperations.CreateTemplate, _subject);
				_biometricClient.BeginPerformTask(task, OnExtractionCompleted, null);
			}
			catch (Exception ex)
			{
				Utils.ShowException(ex);
			}
		}

		private void BtnSaveTemplateClick(object sender, EventArgs e)
		{
			if (saveFileDialog.ShowDialog() == DialogResult.OK)
			{
				try
				{
					File.WriteAllBytes(saveFileDialog.FileName, _subject.GetTemplateBuffer().ToArray());
				}
				catch (Exception ex)
				{
					Utils.ShowException(ex);
				}
			}
		}

		private void BtnSaveVoiceClick(object sender, EventArgs e)
		{
			if (saveVoiceFileDialog.ShowDialog() == DialogResult.OK)
			{
				try
				{
					NVoice voice = _subject.Voices.Last();
					File.WriteAllBytes(saveVoiceFileDialog.FileName, voice.SoundBuffer.Save().ToArray());
				}
				catch (Exception ex)
				{
					Utils.ShowException(ex);
				}
			}
		}

		private void ChbExtractTextDependentCheckedChanged(object sender, EventArgs e)
		{
			_biometricClient.VoicesExtractTextDependentFeatures = chbExtractTextDependent.Checked;
		}

		private void ChbExtractTextIndependentCheckedChanged(object sender, EventArgs e)
		{
			_biometricClient.VoicesExtractTextIndependentFeatures = chbExtractTextIndependent.Checked;
		}

		private void EnrollFromFileVisibleChanged(object sender, EventArgs e)
		{
			if (Visible && _biometricClient != null)
			{
				nudPhraseId.Value = 0;
			}
		}

		#endregion
	}
}

File: java/Samples/Biometrics/VB/SimpleVoicesSampleVB/EnrollFromFile.vb
Imports System
Imports System.IO
Imports System.Linq
Imports System.Windows.Forms
Imports Neurotec.Biometrics
Imports Neurotec.Biometrics.Client

Partial Public Class EnrollFromFile
	Inherits UserControl
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()
	End Sub

#End Region

#Region "Private fields"

	Private _biometricClient As NBiometricClient
	Private _subject As NSubject
	Private _voice As NVoice

#End Region

#Region "Public properties"

	Public Property BiometricClient() As NBiometricClient
		Get
			Return _biometricClient
		End Get
		Set(ByVal value As NBiometricClient)
			_biometricClient = value
		End Set
	End Property

#End Region

#Region "Private methods"

	Private Sub DisableControls()
		lblSoundFile.Text = String.Empty
		lblStatus.Text = String.Empty
		btnExtract.Enabled = False
		btnSaveTemplate.Enabled = False
		btnSaveVoice.Enabled = False
	End Sub

	Private Sub OnExtractionCompleted(ByVal r As IAsyncResult)
		If InvokeRequired Then
			BeginInvoke(New AsyncCallback(AddressOf OnExtractionCompleted), r)
		Else
			Dim task As NBiometricTask = _biometricClient.EndPerformTask(r)
			Dim status As NBiometricStatus = task.Status
			If status = NBiometricStatus.Ok Then
				btnSaveTemplate.Enabled = True
				btnSaveVoice.Enabled = True
				lblStatus.Text = "Template extracted"
			Else
				lblStatus.Text = String.Format("Extraction failed: {0}.", status)
			End If
		End If
	End Sub

#End Region

#Region "Private form events"

	Private Sub EnrollFromFileLoad(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
		DisableControls()
	End Sub

	Private Sub BtnOpenClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnOpen.Click
		_subject = Nothing
		_voice = Nothing

		DisableControls()

		If openFileDialog.ShowDialog() = DialogResult.OK Then
			' Create a subject with voice record
			_voice = New NVoice With {.FileName = openFileDialog.FileName}
			_subject = New NSubject()
			_subject.Voices.Add(_voice)

			lblSoundFile.Text = openFileDialog.FileName
			btnExtract.Enabled = True
		End If
	End Sub

	Private Sub BtnExtractClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnExtract.Click
		btnSaveVoice.Enabled = False
		btnSaveTemplate.Enabled = False
		lblStatus.Text = String.Empty

		If (Not chbExtractTextDependent.Checked) AndAlso (Not chbExtractTextIndependent.Checked) Then
			MessageBox.Show("No features configured to extract", "Invalid options", MessageBoxButtons.OK, MessageBoxIcon.Warning)
			Return
		End If

		Try
			_voice.PhraseId = Convert.ToInt32(nudPhraseId.Value)
			' Do voice extraction and segment voice from audio
			Dim task As NBiometricTask = _biometricClient.CreateTask(NBiometricOperations.Segment Or NBiometricOperations.CreateTemplate, _subject)
			_biometricClient.BeginPerformTask(task, AddressOf OnExtractionCompleted, Nothing)
		Catch ex As Exception
			Utils.ShowException(ex)
		End Try
	End Sub

	Private Sub BtnSaveTemplateClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnSaveTemplate.Click
		If saveFileDialog.ShowDialog() = DialogResult.OK Then
			Try
				File.WriteAllBytes(saveFileDialog.FileName, _subject.GetTemplateBuffer().ToArray())
			Catch ex As Exception
				Utils.ShowException(ex)
			End Try
		End If
	End Sub

	Private Sub BtnSaveVoiceClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnSaveVoice.Click
		If saveVoiceFileDialog.ShowDialog() = DialogResult.OK Then
			Try
				Dim voice As NVoice = _subject.Voices.Last()
				File.WriteAllBytes(saveVoiceFileDialog.FileName, voice.SoundBuffer.Save().ToArray())
			Catch ex As Exception
				Utils.ShowException(ex)
			End Try
		End If
	End Sub

	Private Sub ChbExtractTextDependentCheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles chbExtractTextDependent.CheckedChanged
		If _biometricClient IsNot Nothing Then
			_biometricClient.VoicesExtractTextDependentFeatures = chbExtractTextDependent.Checked
		End If
	End Sub

	Private Sub ChbExtractTextIndependentCheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles chbExtractTextIndependent.CheckedChanged
		If _biometricClient IsNot Nothing Then
			_biometricClient.VoicesExtractTextIndependentFeatures = chbExtractTextIndependent.Checked
		End If
	End Sub

	Private Sub EnrollFromFileVisibleChanged(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.VisibleChanged
		If Visible AndAlso _biometricClient IsNot Nothing Then
			chbExtractTextIndependent.Checked = True
			chbExtractTextDependent.Checked = True
			nudPhraseId.Value = 0
		End If
	End Sub

#End Region
End Class
